sensor/common/heritage/heritage.go

```diff
@@ -0,0 +1,292 @@
+	sensorStart             time.Time
+	sensorVersion           string
+	lastUpdateOfCurrentData time.Time
+
+	maxSize int
+	minSize int
+	maxAge  time.Duration
+
+	// Cache the data from the ConfigMap about the past instances of Sensor
+	cacheIsPopulated atomic.Bool
+	cacheMutex       sync.Mutex
+	cache            []*PastSensor
+}
+
+func NewHeritageManager(ns string, client k8sClient, start time.Time) *Manager {
+	maxSize := env.PastSensorsMaxEntries.IntegerSetting()
+	if maxSize < heritageMinSize {
+		maxSize = heritageMinSize
+		log.Warnf("Value of %s it too low. Applying %d", env.PastSensorsMaxEntries.EnvVar(), heritageMinSize)
+	}
+	return &Manager{
+		cacheIsPopulated: atomic.Bool{},
+		k8sClient:        client,
+		cache:            []*PastSensor{},
+		namespace:        ns,
+		sensorStart:      start,
+		sensorVersion:    version.GetMainVersion(),
+		maxSize:          env.PastSensorsMaxEntries.IntegerSetting(),
+		minSize:          heritageMinSize,
+		maxAge:           heritageMaxAge,
+	}
+}
+
+func (h *Manager) populateCacheFromConfigMap(ctx context.Context) error {
+	data, err := h.readConfigMap(ctx)
+	if err != nil {
+		if apiErrors.IsNotFound(err) {
+			h.cacheIsPopulated.Store(true)
+			log.Debug("No heritage data found. Starting with empty cache")
+			return nil
+		}
+		log.Warnf("Loading data from configMap failed: %v", err)
+		h.cacheIsPopulated.Store(false)
+		return err
+	}
+	log.Infof("Sensor heritage data with %d entries loaded to memory: %s", len(data), pastSensorDataString(data))
+	h.cache = append(h.cache, data...)
+	h.cacheIsPopulated.Store(true)
+	return nil
+}
+
+func (h *Manager) GetData(ctx context.Context) []*PastSensor {
+	h.cacheMutex.Lock()
+	defer h.cacheMutex.Unlock()
+	if h.cacheIsPopulated.Load() {
+		return h.cache
+	}
+	if err := h.populateCacheFromConfigMap(ctx); err != nil {
+		log.Warnf("%v", err)
+	}
+	return h.cache
+}
+
+func (h *Manager) HasCurrentSensorData() bool {
+	return h.currentIP != "" && h.currentContainerID != ""
+}
+
+func (h *Manager) SetCurrentSensorData(currentIP, currentContainerID string) {
+	h.currentIP = currentIP
+	h.currentContainerID = currentContainerID
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+	defer cancel()
+	if err := h.UpsertConfigMap(ctx, time.Now()); err != nil {
+		log.Warnf("Failed to update heritage data in the configMap: %v", err)
+	}
+}
+
+// updateCachedTimestampNoLock updates the timestamp if container ID and IP already exist in the heritage.
+// The size of h.cache is expected to be <10 in most of the cases.
+func (h *Manager) updateCachedTimestampNoLock(now time.Time) bool {
+	for _, entry := range h.cache {
+		if entry.ContainerID == h.currentContainerID && entry.PodIP == h.currentIP {
+			if entry.SensorStart.IsZero() {
+				entry.SensorStart = now
+			}
+			entry.LatestUpdate = now
+			return true
+		}
+	}
+	return false
+}
+
+func (h *Manager) UpsertConfigMap(ctx context.Context, now time.Time) error {
+	if ctx == nil {
```

Comment:
`ctx` must not be nil.

Reply:
Done in [3fa3a90](https://github.com/stackrox/stackrox/pull/15176/commits/3fa3a90d114a61241c148bb3aafe526cfa72ab49)